code/frameworks/angular/src/client/angular-beta/utils/BootstrapLock.ts

```diff
@@ -0,0 +1,50 @@
+/* eslint-disable no-await-in-loop */
+import { ApplicationRef } from '@angular/core';
+
+const queue: Array<() => Promise<void>> = [];
+let isProcessing = false;
+
+/**
+ * Reset compiled components because we often want to compile the same component with
+ * more than one NgModule.
+ */
+const resetCompiledComponents = async () => {
+  try {
+    // Clear global Angular component cache in order to be able to re-render the same component across multiple stories
+    //
+    // References:
+    // https://github.com/angular/angular-cli/blob/master/packages/angular_devkit/build_angular/src/webpack/plugins/hmr/hmr-accept.ts#L50
+    // https://github.com/angular/angular/blob/2ebe2bcb2fe19bf672316b05f15241fd7fd40803/packages/core/src/render3/jit/module.ts#L377-L384
+    const { ɵresetCompiledComponents } = await import('@angular/core');
+    ɵresetCompiledComponents();
+  } catch (e) {
+    /**
+     * noop catch
+     * This means angular removed or modified ɵresetCompiledComponents
+     */
+  }
+};
+
+export const queueBootstrapping = (fn: () => Promise<ApplicationRef>): Promise<ApplicationRef> => {
+  return new Promise<ApplicationRef>((resolve, reject) => {
+    queue.push(() => fn().then(resolve).catch(reject));
+
+    if (!isProcessing) {
+      processQueue();
+    }
+  });
+};
+
+const processQueue = async () => {
+  isProcessing = true;
+
+  while (queue.length > 0) {
+    const bootstrappingFn = queue.shift();
+    if (bootstrappingFn) {
+      await bootstrappingFn();
+      await resetCompiledComponents();
```

Comment:
I haven't tested to see if it would work, but do you think it would make more sense to call this before bootstrapping or keep it after?

The following are my two thoughts:
1. By keeping it after, there may be more consistent usage.
2. If before, it should still prevent conflict with a previous render, but allow usage of not resetting. I don't know if `resetCompiledComponents` can have an effect on lazy loaded modules that aren't part of the initial render. I assume most situations that render multiple stories are dealing more with a documentation scenario, but if `resetCompiledComponents` is hiding an error that would typically happen, would it be better to not call it after? An isolated render in Canvas view could then expose that error, if possible. Of course, then we run into a situation of potential render inconsistency between Canvas and Docs.

Reply:
I agree with your thoughts. I guess we can try it out to place it before? Technically, this might be a breaking change in behavior, but since we have started to work on 8.0.0, we will not release another minor release of Storybook 7.x. I would vote to turn around the logic, as you mentioned.

Reply:
@Marklb Keeping the reset before the actual rendering was [causing issues](https://www.chromatic.com/build?appId=6322ce6af69825592bbb28fc&number=8400). In particular, registered modules like the `CommonModule weren't available. This has lead to styles not being applies since the CommonModule is necessary for the `styles` attribute being applied.